File: models/music.go
package models

import (
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"strconv"
	"strings"

	"ibubble/util"

	"github.com/jinzhu/gorm"
	id3 "github.com/mikkyang/id3-go"
	v2 "github.com/mikkyang/id3-go/v2"
)

// Artist represents an artist
type Artist struct {
	gorm.Model
	Name   string `gorm:"not null"`
	Ruby   string
	Albums []Album
	Songs  []Song
}

// Album represents a music album
type Album struct {
	gorm.Model
	Title      string `gorm:"not null"`
	ArtistID   uint   `gorm:"not null"`
	Year       int    `gorm:"type:smallint unsigned"`
	TrackCount int    `gorm:"type:tinyint unsigned"`
	DiscCount  int    `gorm:"type:tinyint unsigned;default:1"`
	Songs      []Song
}

// Song represents a song
type Song struct {
	gorm.Model
	Filename   string `gorm:"not null"`
	Hash       string `gorm:"unique;not null"`
	Title      string `gorm:"not null"`
	Artist     string
	ArtistID   uint
	AlbumID    uint
	TrackNum   int  `gorm:"type:tinyint unsigned"`
	DiscNum    int  `gorm:"type:tinyint unsigned;default:1"`
	HasArtwork bool `gorm:"not null;default:false"`
	HasLyric   bool `gorm:"not null;default:false"`
	Length     int  `gorm:"type:smallint unsigned;not null"`
	Rate       int  `gorm:"type:tinyint unsigned"`
}

func init() {
	tables := []interface{}{&Artist{}, &Album{}, &Song{}}
	db.AutoMigrate(tables...)
	// db.DropTableIfExists(tables...)
	// db.CreateTable(tables...)
}

// GetSongs returns all songs
func GetSongs() []Song {
	var songs []Song
	db.Find(&songs)
	return songs
}

// ScanAll scans all mp3 files in the music directory
func ScanAll(dir string) error {
	scanErrors := make(map[string]error)

	files, err := getfiles(dir)
	if err != nil {
		return err
	}
	for _, file := range files {
		err := NewSong(file, path.Base(file), false)
		if err != nil {
			scanErrors[file] = err
		}
	}
	for file, err := range scanErrors {
		fmt.Println("error occurred when handling", file)
		fmt.Printf("\t%v\n", err)
	}
	if len(scanErrors) > 0 {
		return errors.New("scan error")
	}
	return nil
}

// NewSong create record of newly added song
func NewSong(filename, originalFileName string, allowOverwrite bool) error {
	conf := util.GetConf()

	if !util.FileExists(filename) {
		return errors.New("file not found")
	}

	artist, album, song, err := loadID3Tags(filename)
	if err != nil {
		return err
	}

	song.Hash = util.Hash([]byte(filename))
	song.Filename = genFilename(artist, album, song, originalFileName)

	// Move
	dst := conf.Storage.Music + "/" + song.Filename
	if !allowOverwrite && util.FileExists(dst) {
		return errors.New("already exists")
	}
	if filename != dst {
		// fmt.Println("a")
		if err = os.MkdirAll(path.Dir(dst), 0755); err != nil {
			return err
		}
		// fmt.Println(util.FileExists(filename))
		if err = os.Rename(filename, dst); err != nil {
			return err
		}
	}

	// Save to DB
	err = db.Where(&Artist{Name: artist.Name}).First(&artist).Error
	if err != nil {
		db.Create(&artist)
	}
	album.ArtistID = artist.ID
	song.ArtistID = artist.ID

	var _album Album
	err = db.Where(&Album{Title: album.Title, ArtistID: album.ArtistID}).First(&_album).Error
	if err != nil {
		db.Create(album)
	} else {
		changed := false
		if _album.Year == 0 && album.Year != 0 {
			_album.Year = album.Year
			changed = true
		}
		if _album.TrackCount == 0 && album.TrackCount != 0 {
			_album.TrackCount = album.TrackCount
			changed = true
		}
		if _album.DiscCount == 0 && album.DiscCount != 0 {
			_album.DiscCount = album.DiscCount
			changed = true
		}
		if changed {
			db.Save(&_album)
		}
	}
	if album.ID != 0 {
		song.AlbumID = album.ID
	} else {
		song.AlbumID = _album.ID
	}

	db.Create(&song)

	return nil
}

func loadID3Tags(filename string) (*Artist, *Album, *Song, error) {
	var artist Artist
	var album Album
	var song Song

	// 初期化. ID3v2.3への変換など. 演奏時間が出力される
	out, err := exec.Command("python", "./scripts/processMP3.py", "init", filename).Output()
	if err != nil {
		return nil, nil, nil, err
	}
	l, err := strconv.Atoi(string(out))
	if err != nil {
		return nil, nil, nil, err
	}
	song.Length = l

	// ID3タグ読み込み
	file, err := id3.Open(filename)

	if err != nil {
		return nil, nil, nil, err
	}
	defer file.Close()

	for _, f := range file.AllFrames() {
		switch f.Id() {
		case "TIT2": // Title
			song.Title = getID3String(&f)
		case "TPE1": // Artist
			song.Artist = getID3String(&f)
		case "TPE2": // Album artist
			artist.Name = getID3String(&f)
		case "TALB": // Album
			album.Title = getID3String(&f)
		case "TYER": // Year
			if y, err := strconv.Atoi(getID3String(&f)); err == nil {
				album.Year = y
			}
		case "TRCK": // Track number
			track := getID3String(&f)
			slice := strings.Split(track, "/")
			if len(slice) == 1 {
				n, err := strconv.Atoi(track)
				if err == nil {
					song.TrackNum = n
				}
			} else { // len(slice) > 1
				n, err := strconv.Atoi(slice[0])
				total, err2 := strconv.Atoi(slice[1])
				if err == nil && err2 == nil {
					song.TrackNum = n
					album.TrackCount = total
				}
			}
		case "TPOS": // Disc number
			track := getID3String(&f)
			slice := strings.Split(track, "/")
			if len(slice) == 1 {
				n, err := strconv.Atoi(track)
				if err == nil {
					song.DiscNum = n
				}
			} else { // len(slice) > 1
				n, err := strconv.Atoi(slice[0])
				total, err2 := strconv.Atoi(slice[1])
				if err == nil && err2 == nil {
					song.DiscNum = n
					album.DiscCount = total
				}
			}
		case "USLT": // Lyrics
			song.HasLyric = true
		case "APIC": // Artwork
			song.HasArtwork = true
		default:
			// nop
		}
	}
	return &artist, &album, &song, nil
}

func getfiles(dir string) ([]string, error) {
	files, err := ioutil.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	var paths []string
	for _, file := range files {
		if file.IsDir() {
			files, err := getfiles(filepath.Join(dir, file.Name()))
			if err != nil {
				return nil, err
			}
			paths = append(paths, files...)
			continue
		}
		paths = append(paths, filepath.Join(dir, file.Name()))
	}

	return paths, nil
}

func genFilename(artist *Artist, album *Album, song *Song, originalName string) string {
	// Set filename
	artistName := escapeFilename(artist.Name)
	albumTitle := escapeFilename(album.Title)
	songTitle := escapeFilename(song.Title)
	if artist.Name != "" {
		if album.Title != "" {
			if song.Title != "" {
				return fmt.Sprintf("%s/%s/%02d %s.mp3", artistName, albumTitle, song.TrackNum, songTitle)
			}
			return fmt.Sprintf("%s/%s/%s", artistName, albumTitle, originalName)
		}
		if song.Title != "" {
			return fmt.Sprintf("%s/Unknown Album/%02d %s.mp3", artistName, song.TrackNum, songTitle)
		}
		return fmt.Sprintf("%s/Unknown Album/%s", artistName, originalName)
	}
	if song.Title != "" {
		return fmt.Sprintf("Unknown Artist/%s.mp3", songTitle)
	}
	return fmt.Sprintf("Unknown Artist/%s", originalName)
}

func escapeFilename(name string) string {
	// 環境によってファイル名に使えない文字を取り除く
	for _, c := range []string{"\\", "/", "|", ":", "*", "?", "\"", "<", ">"} {
		name = strings.Replace(name, c, " ", -1)
	}
	return name
}

func getID3String(f *v2.Framer) string {
	return strings.TrimRight((*f).String(), string(byte(0)))
}

// GetPath returns path to requested file
func GetPath(hash, name string) (string, error) {
	var song Song
	var path string
	err := db.Where("hash = ?", hash).First(&song).Error
	if err != nil {
		return path, err
	}
	if name != escapeFilename(song.Title) {
		return path, errors.New("Not found")
	}

	var artist Artist
	var album Album
	db.Model(&song).Related(&artist)
	db.Model(&song).Related(&album)
	path = fmt.Sprintf("%s/%s", util.GetConf().Storage.Music, genFilename(&artist, &album, &song, name))
	return path, nil
}

File: models/music.rb
require 'digest/md5'
require 'fileutils'
require_relative 'utils/mp3'

class Song < Sequel::Model(:songs)
  many_to_one :album, order: [:year, :title]
  many_to_one :artist, order: [:name]
  many_to_many :playlists

  def self.fix_mp3(filename)
    # deprecated
    # MP3Infoのtagの保存に一部不具合あり
    # ffmpegやchromeで見たときにbitrateがおかしくなり、再生時間が正しく取得できない
    # なのでffmpegで書き直して修正
    tmp = "#{filename}.tmp.mp3"
    out = `ffmpeg -i #{filename.escape_shell} -c copy #{tmp.escape_shell}`
    unless $?.success?
      File.delete(tmp)
      raise "failed: #{out}"
    end
    File.rename(tmp, filename)
  end

  def self.set_tags(filename, data, delete: false)
    return if data.nil?
    tags = MP3.new(filename).tags
    tags.update_to_v24
    tags.delete_tag('TSSE')
    tags.title = data[:title] if data[:title]
    tags.artist = data[:artist] if data[:artist]
    tags.album_artist = data[:album_artist] if data[:album_artist]
    tags.album = data[:album] if data[:album]
    tags.year = data[:year] if data[:year]
    tags.track = data[:track] if data[:track]
    tags.disc = data[:disc] if data[:disc]
    tags.lyrics = data[:lyrics] if data[:lyrics]
    if data[:artwork]
      tags.picture = {
        mime: data[:artwork][:mime],
        data: data[:artwork][:data],
      }
    end
    tags.save
  end

  def self.load_from_tags(tags)
    song = Song.new
    album = Album.new
    artist = Artist.new
    tags.each do |key, val|
      case key
      when 'TIT2' # Title
        song.title = val
      when 'TPE1' # Artist
        song.artist_name = val
      when 'TPE2' # Album artist
        artist.name = val
      when 'TALB' # Album
        album.title = val
      when 'TDRC' # Year
        album.year = val.to_i if val.to_i > 0
      when 'TRCK' # Track number
        values = val.split('/')
        if values.length == 1
          song.track_num = val.to_i if val.to_i > 0
        else # values.length > 1
          n = values[0].to_i
          total = values[1].to_i
          song.track_num = n if n > 0
          album.num_tracks = total if total > 0
        end
      when 'TPOS' # Disc number
        values = val.split('/')
        if values.length == 1
          song.disc_num = val.to_i if val.to_i > 0
        else # values.length > 1
          n = values[0].to_i
          total = values[1].to_i
          song.disc_num = n if n > 0
          album.num_discs = total if total > 0
        end
      when 'USLT' # Lyrics
        song.has_lyric = true
      when 'APIC' # Cover art
        song.has_artwork = true
      end
    end
    [song, album, artist]
  end

  def self.create_from_file(path, filename=nil)
    filename ||= File.basename(path)
    # load id3 tag
    mp3 = MP3.new(path)
    tags = mp3.tags
    tags.update_to_v24

    # load tags
    song, album, artist = load_from_tags(tags)

    unless song.title
      ext = File.extname(filename)
      song.title = File.basename(filename)[%r((.*)#{ext}), 1]
    end

    if artist.name
      song.artist = Artist.first(artist.to_hash) || artist.save
      song.artist_name = artist.name unless song.artist_name
      album.artist_id = song.artist.id
    end

    if album.artist_id and album.title
      song.album = Album.first(album.to_hash.slice(:artist_id, :title))
      if song.album
        [:year, :num_tracks, :num_discs].each{|k|
          song.album[k] = album[k] if song.album[k].nil? and album[k]
        }
      else
        song.album = album.save
      end
    end

    song.length = mp3.length
    song.filename = song.to_filename(filename)
    song

    song.digest = Digest::MD5.file(path).hexdigest[0,8]

    new_filename = "#{CONF.storage.music}/#{song.filename}"
    if File.exist?(new_filename) and Song.first(artist_name: song.artist_name, title: song.title) != nil
      return nil
    end
    if File.realpath(path) != new_filename
      dir = File.dirname(new_filename)
      FileUtils.mkdir_p(dir) unless Dir.exist?(dir)
      File.rename(path, new_filename)
    end
    song.save
  end

  def self.get_path(digest, name)
    song = self.first(digest: digest)
    return if song.nil?
    return if File.basename(song.filename) != name
    "#{CONF.storage.music}/#{song.to_filename}"
  end

  def update_tag(new_tags)
    path = self.to_fullpath
    tags = MP3.new(path).tags
    new_tags.each do |k, v|
      case k
      when 'TYER'
        tags['TDRC'] = v
      else
        tags[k] = v
      end
    end
    tags.save

    self.digest = Digest::MD5.file(path).hexdigest[0,8]

    song, album, artist = self.class.load_from_tags(new_tags)

    if artist.name == nil
      self.artist = nil
    elsif self.artist&.name != artist.name
      self.artist = Artist.first(artist.to_hash) || artist.save
      self.artist_name = artist.name unless self.artist_name or song.artist_name
      if self.album and self.album.title == album.title
        self.album.artist_id = self.artist.id
      end
    end
    album.artist_id = self.artist&.id

    if album.title == nil
      self.album = nil
    elsif album.artist_id == nil
      # nop
    elsif self.album&.title != album.title
      _album = Album.first(album.slice(:artist_id, :title))
      if _album
        self.album = _album
        [:year, :num_tracks, :num_discs].each{|k|
          self.album[k] = album[k] if album[k]
        }
      else
        self.album = album.save
      end
    elsif self.album
      self.album.update(album.to_hash)
    end

    [:title, :artist_name, :track_num, :disc_num].each{|k|
      self[k] = song[k] if song[k]
    }

    if self.filename != self.to_filename
      self.filename = self.to_filename
      new_filename = "#{CONF.storage.music}/#{self.filename}"
      dir = File.dirname(new_filename)
      FileUtils.mkdir_p(dir) unless Dir.exist?(dir)
      File.rename(path, new_filename)
    end

    self.save
    self.update(song.to_hash)
  end

  def fix
    self.class.fix_mp3(self.to_fullpath)
  end

  def to_filename(original_name = nil)
    artist = self.artist&.name&.escape_filename || 'Unknown Artist'
    album = self.album&.title&.escape_filename || 'Unknown Album'
    title = "#{self.title&.escape_filename}.mp3" || original_name || 'Unknown Song.mp3'
    if self.track_num
      "#{artist}/#{album}/%02d #{title}" % self.track_num
    else
      "#{artist}/#{album}/#{title}"
    end
  end

  def to_fullpath
    "#{CONF.storage.music}/#{self.filename}"
  end
end

class Album < Sequel::Model(:albums)
  one_to_many :songs
  many_to_one :artist
end

class Artist < Sequel::Model(:artists)
  one_to_many :songs
  one_to_many :albums
end

class Playlist < Sequel::Model(:playlists)
  many_to_many :songs
end

class PlaylistSong < Sequel::Model(:playlists_songs)
  many_to_one :playlists
  many_to_one :songs
end
